include/wil/cppwinrt_notifiable_module_lock.h

```diff
@@ -0,0 +1,96 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+//! @file
+//! Utilities for implementing OOP COM server with cppwinrt only
+
+#ifndef __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_INCLUDED
+#define __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_INCLUDED
+
+#ifdef WINRT_BASE_H
+#error You must include this header before including any winrt header
+#endif
+
+#ifndef WINRT_CUSTOM_MODULE_LOCK
+#error You must define 'WINRT_CUSTOM_MODULE_LOCK' at the project level
+#endif
+
+#include <atomic>
+#include <cstdint>
+
+namespace wil
+{
+// Adopted from cppwinrt
+struct notifiable_module_lock
+{
+    notifiable_module_lock() = default;
+
+    notifiable_module_lock(uint32_t count) : m_count(count)
+    {
+    }
+
+    uint32_t operator=(uint32_t count) noexcept
+    {
+        return m_count = count;
+    }
+
+    uint32_t operator++() noexcept
+    {
+        return m_count.fetch_add(1, std::memory_order_relaxed) + 1;
+    }
+
+    uint32_t operator--() noexcept
+    {
+        auto const remaining = m_count.fetch_sub(1, std::memory_order_release) - 1;
+
+        if (remaining == 0)
+        {
+            std::atomic_thread_fence(std::memory_order_acquire);
+            notifier();
+        }
+        else if (remaining < 0)
+        {
+            abort();
+        }
+
+        return remaining;
+    }
+
+    operator uint32_t() const noexcept
+    {
+        return m_count;
+    }
+
+    template <typename Func>
+    static void set_notifier(Func&& func)
+    {
+        notifier = std::forward<Func>(func);
+    }
+
+private:
+    std::atomic<int32_t> m_count;

```

Comment:
```suggestion
    std::atomic<int32_t> m_count{0};
```
It's been a while, but last time I used `std::atomic`, the default constructor doesn't initialize to zero. I think what you currently have is working because it's a `static` variable, which should get zero initialized.

Reply:
Looks like the construction is trivial until C++20, at which point it gets initialized with `T()`. Does C++/WinRT still support C++17?

Reply:
C++/WinRT requires C++17. Think it's best to initialize to 0 nonetheless.

Reply:
Btw this is a direct copy of the lock implementation in cppwinrt. You might want to discuss internally whether that should be fixed/file an issue there.